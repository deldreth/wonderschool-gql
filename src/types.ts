import { InMemoryCache } from 'apollo-cache-inmemory';
import { DateTime } from 'luxon';

export interface Task {
  id: string;
  Group: Group;
  task: string;
  dependencyIds: number[];
  completedAt?: DateTime;
}

export interface Group {
  id: string;
  name: string;
}

export interface StateCache {
  cache: InMemoryCache;
}
