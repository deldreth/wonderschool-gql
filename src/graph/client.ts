import { InMemoryCache } from 'apollo-cache-inmemory';
import ApolloClient from 'apollo-client';
import { ApolloLink } from 'apollo-link';
import { HttpLink } from 'apollo-link-http';

const client = new ApolloClient( {
  cache: new InMemoryCache(),
  link: ApolloLink.from( [
    new HttpLink( {
      uri: 'http://localhost:3000',
    } ),
  ] ),
} );

export default client;
