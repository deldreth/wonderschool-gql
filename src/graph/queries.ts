import gql from 'graphql-tag';
import { graphql } from 'react-apollo';

export const ALL_TASKS_QUERY = gql`
  {
    allTasks {
      id
      completedAt
      dependencyIds
      Group {
        id
        name
      }
    }
  }
`;

export const ALL_GROUPS_QUERY = gql`
  {
    allGroups {
      id
      name
    }
  }
`;

export const GROUP_NAME_QUERY = gql`
  query groupName( $id: ID! ) {
    Group( id: $id ) {
      name
    }
  }
`;

export const GROUP_TASKS_QUERY = gql`
  query allTasksByGroup($group: ID!) {
    allTasks(filter: { group_id: $group }) {
      id
      task
      completedAt
      dependencyIds
    }
  }
`;
