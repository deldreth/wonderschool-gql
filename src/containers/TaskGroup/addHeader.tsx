import React, { SyntheticEvent } from 'react';

import gql from 'graphql-tag';
import { compose, withStateHandlers } from 'recompose';
import styled from 'styled-components';

import { ADD_GROUP_MUTATION } from 'app/graph/mutations';
import withMutation from 'app/graph/withMutation';

import { faPlus, faTimes } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import { Icon, ListContainer, ListItem, Loading } from 'app/styles';
import { Keyboard } from 'puppeteer';

interface InjectedProps {
  adding: boolean;
  newGroup: string;
  isAdding: ( value: boolean ) => void;
  updateNewGroup: ( value: string ) => void;
  createGroup: ( variables: any ) => void;
}

interface ExternalProps {
  nextId: number;
}

type Props = ExternalProps & InjectedProps;

function AddHeader ( { nextId, adding, updateNewGroup, createGroup, isAdding, newGroup }: Props ) {
  if ( adding ) {
    return (
      <ListItem variant="header">
        <input
          type="text"
          onChange={ ( event: any ) => updateNewGroup( event.target.value ) }
          onKeyPress={ ( event: any ) => {
            if ( event.key === 'Enter' ) {
              createGroup( { variables: {
                id: nextId,
                name: newGroup,
              } } ); 
            }
          } }
          autoFocus/>

        <div className="icon" onClick={ () => {
          if ( newGroup !== '' ) {
            createGroup( { variables: {
              id: nextId,
              name: newGroup,
            } } ); 
          }
        } }>
          <FontAwesomeIcon icon={ faPlus } />
        </div>

        <div className="icon" onClick={ () => isAdding( false ) }>
          <FontAwesomeIcon icon={ faTimes } />
        </div>
      </ListItem>
    );
  } else {
    return (
      <ListItem variant="header">
        Things To Do
        <div onClick={ () => isAdding( true ) }>
          <FontAwesomeIcon icon={ faPlus } />
        </div>
      </ListItem>
    );
  }
}

export default compose<InjectedProps, ExternalProps>(
  withMutation( ADD_GROUP_MUTATION, 'createGroup' ),
  withStateHandlers(
    () => ( {
      adding: false,
      newGroup: '',
    } ),
    {
      isAdding: ( { adding } ) => ( value ) => ( {
        adding: value,
      } ),
      updateNewGroup: ( { newGroup } ) => ( value ) => ( {
        newGroup: value,
      } ),
    },
  ),
)( AddHeader );
