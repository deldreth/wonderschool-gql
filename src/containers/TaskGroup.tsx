import React from 'react';

import gql from 'graphql-tag';
import { Mutation, Query } from 'react-apollo';
import { RouteComponentProps, withRouter } from 'react-router-dom';
import { compose, withHandlers, withProps } from 'recompose';
import styled from 'styled-components';

import { faCaretRight } from '@fortawesome/free-solid-svg-icons/faCaretRight';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import { ADD_GROUP_MUTATION } from 'app/graph/mutations';
import { ALL_GROUPS_QUERY, ALL_TASKS_QUERY, GROUP_TASKS_QUERY } from 'app/graph/queries';
import withMutation from 'app/graph/withMutation';
import withQuery from 'app/graph/withQuery';

import Icon from 'app/components/Icon';
import InputHeader from 'app/components/InputHeader';
import { ListContainer, ListItem, Loading } from 'app/components/Styles';
import { Group, Task } from 'app/types';

interface ExternalProps {
}

export interface InjectedProps {
  allTasks: Task[];
  loading: boolean;
  routeTo: ( path: string ) => () => void;
  aggregates: {
    [ group: string ]: {
      completed: number,
      total: number,
    },
  };
}

export type Props = ExternalProps & InjectedProps & RouteComponentProps<any>;

function TaskGroup ( { routeTo, aggregates }: Props ) {
  return (
    <ListContainer>
      <Query query={ ALL_GROUPS_QUERY } pollInterval={ 500 }>
        {
          ( { loading, data: { allGroups }, refetch } ) => {
            if ( loading ) {
              return <Loading>...</Loading>;
            }

            return [
              <Mutation
                key="group-add-header"
                mutation={ ADD_GROUP_MUTATION }>
                { ( createGroup ) => (
                  <InputHeader
                    createMutationHandler={ createGroup }
                    nextId={ allGroups.length + 1 }
                    postAdd={ refetch }/>
                ) }
              </Mutation>,
              allGroups.map( ( group: Group ) => (
                <TaskGroupItem key={ group.id }
                  onClick={ routeTo( `/group/${group.id}` ) }>
                  <Icon icon={ faCaretRight } />
      
                  <TaskGroupItemText>
                    { group.name }
      
                    <Aggregate
                      aggregates={ aggregates }
                      groupName={ group.name }/>
                  </TaskGroupItemText>
                </TaskGroupItem>
              ) ),
            ];
          }
        }
      </Query>
    </ListContainer>
  );
}

const Aggregate = ( { aggregates, groupName }: any ) => {
  if ( aggregates && aggregates[ groupName ] ) {
    return (
      <TaskGroupItemAggregate>
        { aggregates[ groupName ].completed } of { aggregates[ groupName ].total } tasks complete
      </TaskGroupItemAggregate>
    );
  }

  return null;
};

export default compose<InjectedProps, ExternalProps>(
  withRouter,
  withQuery( ALL_TASKS_QUERY ),
  withProps( ( { allTasks, loading }: Props ) => {
    if ( loading ) {
      return {};
    }

    const aggregates: Props['aggregates'] = {};
    allTasks.forEach( task => {    
      if ( !aggregates.hasOwnProperty( task.Group.name ) ) {
        aggregates[ task.Group.name ] = {
          completed: 0,
          total: 0,
        };
      }
      
      aggregates[ task.Group.name ].completed += task.completedAt ? 1 : 0;
      aggregates[ task.Group.name ].total += 1;
    } );
    
    return { aggregates };
  } ),
  withHandlers( {
    routeTo: ( { history }: Props ) => ( value: string )  => () => {
      history.push( value );
    },
  } ),
)( TaskGroup );

const TaskGroupItem = ListItem.extend`
  &:hover {
    cursor: pointer;
  }
`;

const TaskGroupItemText = styled.div`
`;

const TaskGroupItemAggregate = styled.div`
  padding: 4px 0px 0px 0px;
  color: #AAAAAA;
`;
