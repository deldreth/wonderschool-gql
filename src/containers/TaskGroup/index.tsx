import React from 'react';

import gql from 'graphql-tag';
import { Query } from 'react-apollo';
import { RouteComponentProps, withRouter } from 'react-router-dom';
import { compose, withHandlers, withProps, withStateHandlers } from 'recompose';
import styled from 'styled-components';

import { ADD_GROUP_MUTATION } from 'app/graph/mutations';
import { ALL_GROUPS_QUERY, ALL_TASKS_QUERY, GROUP_TASKS_QUERY } from 'app/graph/queries';
import withMutation from 'app/graph/withMutation';
import withQuery from 'app/graph/withQuery';

import { Icon, ListContainer, ListItem, Loading } from 'app/styles';
import { Group, Task } from 'app/types';

import AddHeader from './addHeader';

interface ExternalProps {
}

export interface InjectedProps {
  allTasks: Task[];
  loading: boolean;
  routeTo: ( path: string ) => () => void;
  aggregates: {
    [ group: string ]: {
      completed: number,
      total: number,
    },
  };
}

export type Props = ExternalProps & InjectedProps & RouteComponentProps<any>;

function TaskGroup ( { routeTo, aggregates }: Props ) {
  return (
    <ListContainer>
      <Query query={ ALL_GROUPS_QUERY } pollInterval={ 500 }>
        {
          ( { loading, data: { allGroups } } ) => {
            if ( loading ) {
              return <Loading>...</Loading>;
            }

            return [
              <AddHeader
                key="group-add-header"
                nextId={ allGroups.length + 1 } />,
              allGroups.map( ( group: Group ) => (
                <TaskGroupItem key={ group.id }
                  onClick={ routeTo( `/group/${group.id}` ) }>
                  <Icon>
                    <img src={ require( 'static/Group.svg' ) }/>
                  </Icon>
      
                  <TaskGroupItemText>
                    { group.name }
      
                    { ( aggregates && aggregates[group.name] ) ? 
                      <Aggregate agg={ aggregates[group.name] }/>
                    : null }
                  </TaskGroupItemText>
                </TaskGroupItem>
              ) ),
            ];
          }
        }
      </Query>
    </ListContainer>
  );
}

function Aggregate ( { agg }: any ) {
  return (
    <TaskGroupItemAggregate>
      { agg.completed } OF { agg.total } TASKS COMPLETE
    </TaskGroupItemAggregate>
  );
} 

export default compose<InjectedProps, ExternalProps>(
  withRouter,
  withQuery( ALL_TASKS_QUERY ),
  withProps( ( { allTasks, loading }: Props ) => {
    if ( loading ) {
      return {};
    }

    const aggregates: Props['aggregates'] = {};
    allTasks.forEach( task => {    
      if ( !aggregates.hasOwnProperty( task.Group.name ) ) {
        aggregates[ task.Group.name ] = {
          completed: 0,
          total: 0,
        };
      }
      
      aggregates[ task.Group.name ].completed += task.completedAt ? 1 : 0;
      aggregates[ task.Group.name ].total += 1;
    } );
    
    return { aggregates };
  } ),
  withHandlers( {
    routeTo: ( { history }: Props ) => ( value: string )  => () => {
      history.push( value );
    },
  } ),
)( TaskGroup );

const TaskGroupItem = ListItem.extend`
  &:hover {
    cursor: pointer;
  }
`;

const TaskGroupItemText = styled.div`
`;

const TaskGroupItemAggregate = styled.div`
  padding: 4px 0px 0px 0px;
  color: #AAAAAA;
`;
