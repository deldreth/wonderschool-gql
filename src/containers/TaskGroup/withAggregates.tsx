import React from 'react';

import { Query } from 'react-apollo';
import compose from 'recompose/compose';

import { InjectedProps } from 'app/containers/TaskGroup';

export interface Aggregate {
  [ group: string ]: {
    completed: number,
    total: number,
  };
}

/**
 * HOC to generate group aggregate completes and totals
 */

export default <TPropsInjected extends InjectedProps>
  ( Component: React.ComponentType<TPropsInjected> ) => 
    ( props: TPropsInjected ) => {
      if ( props.loading ) {
        return <Component { ...props }/>;
      }

      const aggregates: Aggregate = {};
      props.allTasks.forEach( task => {    
        if ( !aggregates.hasOwnProperty( task.group ) ) {
          aggregates[ task.group ] = {
            completed: 0,
            total: 0,
          };
        }
        
        aggregates[ task.group ].completed += task.completedAt ? 1 : 0;
        aggregates[ task.group ].total += 1;
      } );

      return <Component { ...props } aggregates={ aggregates } />;
    };
