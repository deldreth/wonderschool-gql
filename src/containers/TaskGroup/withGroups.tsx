import React from 'react';

import { Query } from 'react-apollo';
import compose from 'recompose/compose';

import { InjectedProps } from 'app/containers/TaskGroup';

/**
 * HOC to generate task groups
 */

export default <TPropsInjected extends InjectedProps>
  ( Component: React.ComponentType<TPropsInjected> ) => 
    ( props: TPropsInjected ) => {
      if ( props.loading ) {
        return <Component { ...props }/>;
      }

      const groups: string[] = [];
      props.allTasks.forEach( task => {
        if ( !groups.includes( task.group ) ) {
          groups.push( task.group );
        }
      } );

      return <Component { ...props } groups={ groups } />;
    };
