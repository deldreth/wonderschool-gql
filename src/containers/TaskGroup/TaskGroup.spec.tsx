import { shallow } from 'enzyme';
import React from 'react';
import { MockedProvider } from 'react-apollo/test-utils';

import { ALL_TASKS_QUERY } from 'app/graph/queries';
import { Task as TaskType } from 'app/types';
import TaskGroup from './index';

it( 'should render a TaskGroup component', async () => {
  const mocks: any = [ {
    request: { query: ALL_TASKS_QUERY },
    result: {
      data: {
        allTasks: [
          {
            id: '1',
            group: 'Purchases',
            task: 'Go to the bank',
            dependencyIds: [1],
            completedAt: null,
          },
          {
            id: '2',
            group: 'Airplanes',
            task: 'Fly One',
            dependencyIds: [1],
            completedAt: null,
          },
        ],
      },
    },
  } ];

  const groups = [ 'Purchases', 'Airplanes' ];

  const wrapper = shallow(
    <MockedProvider mocks={ mocks }>
      <TaskGroup />
    </MockedProvider>,
  );

  expect( wrapper ).toMatchSnapshot();
} );
