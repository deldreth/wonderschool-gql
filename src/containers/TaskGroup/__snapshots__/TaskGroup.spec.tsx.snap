// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a TaskGroup component 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MockedProvider
    addTypename={true}
    mocks={
      Array [
        Object {
          "request": Object {
            "query": Object {
              "definitions": Array [
                Object {
                  "directives": Array [],
                  "kind": "OperationDefinition",
                  "name": undefined,
                  "operation": "query",
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "allTasks",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "id",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "completedAt",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "dependencyIds",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "Group",
                              },
                              "selectionSet": Object {
                                "kind": "SelectionSet",
                                "selections": Array [
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "id",
                                    },
                                    "selectionSet": undefined,
                                  },
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "name",
                                    },
                                    "selectionSet": undefined,
                                  },
                                ],
                              },
                            },
                          ],
                        },
                      },
                    ],
                  },
                  "variableDefinitions": Array [],
                },
              ],
              "kind": "Document",
              "loc": Object {
                "end": 123,
                "start": 0,
              },
            },
          },
          "result": Object {
            "data": Object {
              "allTasks": Array [
                Object {
                  "completedAt": null,
                  "dependencyIds": Array [
                    1,
                  ],
                  "group": "Purchases",
                  "id": "1",
                  "task": "Go to the bank",
                },
                Object {
                  "completedAt": null,
                  "dependencyIds": Array [
                    1,
                  ],
                  "group": "Airplanes",
                  "id": "2",
                  "task": "Fly One",
                },
              ],
            },
          },
        },
      ]
    }
  >
    <withRouter() />
  </MockedProvider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <withRouter() />,
      "client": ApolloClient {
        "cache": InMemoryCache {
          "addTypename": true,
          "config": Object {
            "addTypename": true,
            "dataIdFromObject": [Function],
            "fragmentMatcher": HeuristicFragmentMatcher {},
            "storeFactory": [Function],
          },
          "data": ObjectCache {
            "data": Object {},
          },
          "optimistic": Array [],
          "silenceBroadcast": false,
          "watches": Array [],
        },
        "defaultOptions": Object {},
        "disableNetworkFetches": false,
        "link": ApolloLink {
          "request": [Function],
        },
        "mutate": [Function],
        "query": [Function],
        "queryDeduplication": true,
        "reFetchObservableQueries": [Function],
        "resetStore": [Function],
        "resetStoreCallbacks": Array [],
        "ssrMode": false,
        "store": DataStore {
          "cache": InMemoryCache {
            "addTypename": true,
            "config": Object {
              "addTypename": true,
              "dataIdFromObject": [Function],
              "fragmentMatcher": HeuristicFragmentMatcher {},
              "storeFactory": [Function],
            },
            "data": ObjectCache {
              "data": Object {},
            },
            "optimistic": Array [],
            "silenceBroadcast": false,
            "watches": Array [],
          },
        },
        "version": "2.3.1",
        "watchQuery": [Function],
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <withRouter() />,
        "client": ApolloClient {
          "cache": InMemoryCache {
            "addTypename": true,
            "config": Object {
              "addTypename": true,
              "dataIdFromObject": [Function],
              "fragmentMatcher": HeuristicFragmentMatcher {},
              "storeFactory": [Function],
            },
            "data": ObjectCache {
              "data": Object {},
            },
            "optimistic": Array [],
            "silenceBroadcast": false,
            "watches": Array [],
          },
          "defaultOptions": Object {},
          "disableNetworkFetches": false,
          "link": ApolloLink {
            "request": [Function],
          },
          "mutate": [Function],
          "query": [Function],
          "queryDeduplication": true,
          "reFetchObservableQueries": [Function],
          "resetStore": [Function],
          "resetStoreCallbacks": Array [],
          "ssrMode": false,
          "store": DataStore {
            "cache": InMemoryCache {
              "addTypename": true,
              "config": Object {
                "addTypename": true,
                "dataIdFromObject": [Function],
                "fragmentMatcher": HeuristicFragmentMatcher {},
                "storeFactory": [Function],
              },
              "data": ObjectCache {
                "data": Object {},
              },
              "optimistic": Array [],
              "silenceBroadcast": false,
              "watches": Array [],
            },
          },
          "version": "2.3.1",
          "watchQuery": [Function],
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
